### DISCUSSION TITLE
osu!lazer not using VAAPI decoder on Linux with Intel APU

### DISCUSSION THREAD

**Discussion by elsandosgrande** [ppy/osu#16567] [Q&A]
My laptop — Dell Inspiron 3793 — has an Intel APU — Intel i5-1035G1 — as well as an NVIDIA dGPU — NVIDIA GeForce MX230 — and runs Ubuntu 21.10. When playing a beatmap with video, osu!lazer doesn't try to use the APU for video decoding through VA-API and instead tries to use the NVIDIA dGPU through VDPAU and fails because the dGPU is not equipped with any video ASICs and it then proceeds to try to use the CUDA hardware and fails according to the terminal messages, but it eventually succeeds with CUDA according to the runtime log and `nvidia-smi`.

While yes, the video does play just fine, it would be more power-efficient to use the APU's video ASICs through VA-API than to initialise the dGPU and use CUDA for hardware decoding; my laptop would also have more thermal headroom thanks to the cooling on this laptop being subpar. Are there any practical reasons for preventing VA-API from being utilised by osu!lazer? Firefox and VLC seem to work just fine with VA-API and the game runs on the APU anyways, so I can't really see any reason not to use VA-API when available.

<details>

<summary>Terminal messages</summary>

<pre>
<code>$ ./Downloads/osu.AppImage
Invalid return value 0 for stream protocol
Invalid return value 0 for stream protocol
GPU at BusId 0x1 doesn't have a supported video decoder
[AVHWDeviceContext @ 0x7f8a20aaf940] VDPAU device creation on X11 display :0 failed.
[mpeg4_cuvid @ 0x7f8a20aaad80] Codec mpeg4_cuvid is not supported.
[mpeg4 @ 0x7f8a21281bc0] Hardware is lacking required capabilities
[mpeg4 @ 0x7f8a21281bc0] Failed setup for format cuda: hwaccel initialisation returned error.
[…]</code>
</pre>

</details>

[runtime.log](https://github.com/ppy/osu/files/7919087/runtime.log)

**Comment by bdach** [comment]
VAAPI is actually supported, it's just that the default affinity towards that codec is lower than VDPAU's and CUDA's. see: https://github.com/ppy/osu-framework/blob/554f1f24c1c048849819caeffc11af62428a8e45/osu.Framework/Graphics/Video/AVHWDeviceTypePerformanceComparer.cs#L12-L26

because CUDA initialised successfully, VAAPI never was even tried (because it technically didn't have to be).

i'm not very well versed in video stuff so it's difficult for me to opine whether that default affinity is correct, but that's how things currently stand. we can tweak the affinities if there are valid global adjustments to make, but if the "correct" choice of codec is highly situational and depends on hardware then honestly this is something that i'd rather users had to live with.

i'd be against exposing a user-facing setting to change this, because users commonly misuse low-level settings, believing claims that aren't really substantiated by anything.

**Reply by elsandosgrande** [reply]
After taking a cursory glance at https://github.com/ppy/osu-framework/pull/4839, it seems like it was essentially an arbitrary decision, or an assumption that NVIDIA is always the best choice. 

It's been a while since I've looked at any Firefox VA-API bugs, but I recall some Firefox developer saying that DMA-BUF is used in Firefox on Linux in part to avoid copying frames from GPU memory to CPU memory and back to GPU memory if the GPU that's rendering Firefox is the same as the one that's decoding the frames, which is always the case, since the performance is pretty much the same as just using the CPU for decoding if the frames have to make a round trip; I vaguely recall there being an equivalent API for NVIDIA, at least for CUDA, but only if the NVIDIA GPU is also rendering osu!lazer. Please correct me if I'm wrong, but the messages in the pull request above give me the impression that hardware-accelerated video decoding is used without regard to how many memory copy operations are needed to get the decoded frames to the texture reserved for them in osu!lazer, though that is kind of tangential to what I'm asking for here.

**My proposition**: Have the GPU that's used for video decoding be the same as the GPU used for rendering osu!lazer.

---

### Side note
> […] but if the "correct" choice of codec is highly situational and depends on hardware then honestly this is something that i'd rather users had to live with.

It's not about codecs though, it's about which device is used for video decoding. That said, falling back to software decoding if a given codec is not supported seems to be a fine compromise, since I don't imagine many beatmaps having codecs outside of the following:

<details>

<summary>vainfo</summary>

<pre>
<code>$ vainfo
libva info: VA-API version 1.12.0
libva info: Trying to open /usr/lib/x86_64-linux-gnu/dri/iHD_drv_video.so
libva info: Found init function __vaDriverInit_1_12
libva info: va_openDriver() returns 0
vainfo: VA-API version: 1.12 (libva 2.10.0)
vainfo: Driver version: Intel iHD driver for Intel(R) Gen Graphics - 21.3.3 ()
vainfo: Supported profile and entrypoints
      VAProfileNone                   :	VAEntrypointVideoProc
      VAProfileNone                   :	VAEntrypointStats
      VAProfileMPEG2Simple            :	VAEntrypointVLD
      VAProfileMPEG2Main              :	VAEntrypointVLD
      VAProfileH264Main               :	VAEntrypointVLD
      VAProfileH264Main               :	VAEntrypointEncSliceLP
      VAProfileH264High               :	VAEntrypointVLD
      VAProfileH264High               :	VAEntrypointEncSliceLP
      VAProfileJPEGBaseline           :	VAEntrypointVLD
      VAProfileJPEGBaseline           :	VAEntrypointEncPicture
      VAProfileH264ConstrainedBaseline:	VAEntrypointVLD
      VAProfileH264ConstrainedBaseline:	VAEntrypointEncSliceLP
      VAProfileVP8Version0_3          :	VAEntrypointVLD
      VAProfileHEVCMain               :	VAEntrypointVLD
      VAProfileHEVCMain               :	VAEntrypointEncSliceLP
      VAProfileHEVCMain10             :	VAEntrypointVLD
      VAProfileHEVCMain10             :	VAEntrypointEncSliceLP
      VAProfileVP9Profile0            :	VAEntrypointVLD
      VAProfileVP9Profile0            :	VAEntrypointEncSliceLP
      VAProfileVP9Profile1            :	VAEntrypointVLD
      VAProfileVP9Profile1            :	VAEntrypointEncSliceLP
      VAProfileVP9Profile2            :	VAEntrypointVLD
      VAProfileVP9Profile2            :	VAEntrypointEncSliceLP
      VAProfileVP9Profile3            :	VAEntrypointVLD
      VAProfileVP9Profile3            :	VAEntrypointEncSliceLP
      VAProfileHEVCMain422_10         :	VAEntrypointVLD
      VAProfileHEVCMain444            :	VAEntrypointVLD
      VAProfileHEVCMain444            :	VAEntrypointEncSliceLP
      VAProfileHEVCMain444_10         :	VAEntrypointVLD
      VAProfileHEVCMain444_10         :	VAEntrypointEncSliceLP
[Note: While my APU supports a bit more than this, that's only with Intel's non-free driver, `intel-media-va-driver-non-free`]</code>
</pre>

</details>

## Edit

While I have yet to find the original comment on Bugzilla that refers to the whole copying-to-system-memory thing, I have found a different comment that briefly mentions the same thing: https://bugzilla.mozilla.org/show_bug.cgi?id=1738545#c2

![image](https://user-images.githubusercontent.com/40576902/150656253-a33ebbd1-48b1-4452-977b-ce8e2cbf3b0b.png)

## Edit two

https://bugzilla.mozilla.org/show_bug.cgi?id=1738545#c5

![image](https://user-images.githubusercontent.com/40576902/150656344-98a17c9c-6467-463d-b12e-20c5d6486a14.png)

https://bugzilla.mozilla.org/show_bug.cgi?id=1738545#c8

![image](https://user-images.githubusercontent.com/40576902/150656356-d6d3b22b-a14c-4c07-a0e2-c13774c351f3.png)
 
I'd look for references to Chrome stuff, but I'm not that familiar with the state of its VA-API support 😅

**Reply by bdach** [reply]
> Please correct me if I'm wrong, but the messages in the pull request above give me the impression that hardware-accelerated video decoding is used without regard to how many memory copy operations are needed to get the decoded frames to the texture reserved for them

there may be work remaining there yes. adding support for hardware decoding in the mentioned PR was an incremental step. patches to framework welcome, but it's not generally a pressing issue of high priority.

> My proposition: Have the GPU that's used for video decoding be the same as the GPU used for rendering osu!lazer.

not sure this will happen if FFmpeg doesn't already expose a way to find this out. not sure what else we'd use to reliably determine if it's the same gpu in a cross-platform manner.

> That said, falling back to software decoding if a given codec is not supported seems to be a fine compromise

this will already happen, with a few work items still left to address (https://github.com/ppy/osu-framework/issues/4857).

**Reply by elsandosgrande** [reply]
> patches to framework welcome […]

I've just started learning C++ and Rust this month, so I'm not sure if I'll be able to contribute any code this year, especially given that I also have university to spend time on and that osu!lazer seems to be usign C# instead of C++ 😅 

> […] it's not generally a pressing issue of high priority.

Yeah, I figured that given that nothing's horribly broken here.

> not sure this will happen if FFmpeg doesn't already expose a way to find this out. not sure what else we'd use to reliably determine if it's the same gpu in a cross-platform manner.

I don't think that FFmpeg can do that, even in theory. [This is what Firefox does on Linux](https://bugzilla.mozilla.org/show_bug.cgi?id=1588904), but I'm still looking for equivalent Windows and macOS bugs to pick the mechanism(s) apart.

---

OK, so what DMA-BUF is used for is called “buffer sharing”: https://bugzilla.mozilla.org/show_bug.cgi?id=1588904#c23

![image](https://user-images.githubusercontent.com/40576902/150658353-7c6b090f-a75b-445f-b675-82b7d3d119af.png)

In hindsight, that makes perfect sense, but I'm not sure why it didn't occur to me before that that's called buffer sharing 😅

(Likely) relevant kernel documentation which I just stumbled across: https://www.kernel.org/doc/html/latest/driver-api/dma-buf.html

---

### Post Scriptum

Seeing as osu!lazer runs like a dream on my Core i5 APU, I presume that it also runs pretty well on Intel and AMD APUs from before 2019, especially Core i7 (equivalent) ones, so I'd wager that there's no need for the project to delve into this quagmire: https://bugzilla.mozilla.org/show_bug.cgi?id=1738545#c15

![image](https://user-images.githubusercontent.com/40576902/150659324-b01f33ae-cfe1-410d-9698-a62071f541ba.png)
